Refactor DatabaseInterface.php

kamil-tekiela opened the pull request:
This PR improves code readability, performance and stability of DatabaseInterface.php. These are rather minor changes with the exception of adding a new internal method to simplify the code. 

For the explanation behind each change see commit descriptions.

kamil-tekiela committed:
Use str_starts_with()

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove redundant !empty()

!empty() on a declared variable is redundant.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Replace loop with array_column()

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Simpify assignment by removing temporary variable

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove redundant if statements and array_merge

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove temporary variable

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Unindent code

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Add helper method fetchByMode()

This is done to avoid calling methods through variables. It is the same
in terms of performance.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Move if statements to where they are relevant

I had a choice to either move the while loop out of the if statements
or move the if statements inside. I decided it made more sense to move
them inside.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Add NULL to phpdoc for the two parameters that can be null

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Simplify return statement

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Improve performance and simplify code

Since we are always using STORE mode then we can use data_seek() method.
This will allow us to get rid of the loop and use fetchByMode() method.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Inline value defaulting

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
[bugfix] getColumnNames() should return an array of strings

There were multiple issues here. If we are selecting only a certain
column from the SQL then let's use the handy function instead of
array_keys. We will never get anything other than an array so checking
if it is not an array makes no sense. Returning null only to cast it to
an array is pointless. If really necessary, we could still check for an
empty array. The ColumnController returned the value without even
casting it to an array. I am not sure how jQuery/JavaScript dealt with it.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Update psalm-baseline.xml

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Refactor getColumnMapFromSql and add type hint

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Refactor getVirtualTables

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove redundant loop

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Remove redundant usage of freeResult()

If there was no chance for the object to leak the scope and it is the
end of the scope then there's no reason to call this method.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
Replace dynamic variable with declare one

It's always the same name. If we were creating more than one, we could
use associative array, but in this case there seems to be absolutely no
need for any of this. Dynamic variables make it harder to debug the code.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

williamdes commented on libraries/classes/Database/CentralColumns.php:
```diff
@@ -900,18 +902,9 @@ public function getColumnsNotInCentralList(string $db, string $table): array
-        }
-
-        return $columns;
+        return array_values(array_diff($columnNames, $existingColumns));
```
This is a bit hard to read, could you add a comment about what it does ?

kamil-tekiela replied:
I'm am of opinion that clean code should be self-explanatory and for this reason I prefer to keep code comments to the minimum. I am a little surprised that this simplification is still confusing. It does exactly what the method name says in 3 lines of code. There is one line `selectDb()` that is confusing to me and I can't explain its purpose. 

Nonetheless, I have added a comment right above `return` to make it as clear as possible that this function returns a simple list of column names except the ones in central list of columns. Please let me know if this makes things any clearer. 

P.S. The method `getFromTable()` is confusing, but that's a separate topic not related to this PR.

williamdes replied:
It's very cool with the comment, that helps the brain clear the process between what values are kept from the diff

williamdes commented on libraries/classes/Plugins/Export/ExportMediawiki.php:
```diff
@@ -299,7 +299,7 @@ public function exportData(
 
             // Add column names as table headers
-            if ($column_names !== null) {
+            if ($column_names) {
```
maybe use count or `!= []` ?

mauriciofauth commented on libraries/classes/DatabaseInterface.php:
```diff
@@ -1306,6 +1260,25 @@ private function fetchValueOrValueByIndex($row, $value)
+     * returns array of rows with numeric or associative keys
+     *
+     * @param object $result result set identifier
+     * @param int    $mode   either self::FETCH_ROW or self::FETCH_ASSOC
```
```suggestion
     * @param int    $mode   either self::FETCH_ROW or self::FETCH_ASSOC
     * @psalm-param self::FETCH_ROW|self::FETCH_ASSOC $mode
```

kamil-tekiela replied:
Thanks for this suggestion. I have thought about it and decided that the solution I came up with was too limited in scope. I decided to rewrite this commit and change the parameter type. I moved the constants to the interface and refactored `fetchSingleRow()`. I replaced literal strings with constants. Hopefully this makes the code cleaner and easier to understand.

kamil-tekiela committed:
Remove redundant check

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

williamdes commented on libraries/classes/DatabaseInterface.php:
```diff
@@ -665,7 +656,7 @@ public function getDatabasesFull(
             // get table information from information_schema
             $sqlWhereSchema = '';
-            if (! empty($database)) {
+            if ($database) {
```
```suggestion
            if ($database !== null) {
```

kamil-tekiela replied:
I loosened the condition per your suggestion. Thanks.

williamdes commented on libraries/classes/DatabaseInterface.php:
```diff
@@ -786,24 +777,20 @@ public function getColumnMapFromSql(string $sql_query, array $view_columns = [])
-
-        $nbFields = count($meta);
-        if ($nbFields <= 0) {
+        if (! $meta) {
```
The previous syntax was not better and more strict ?

kamil-tekiela replied:
I admit, this is one of the places where I took the liberty of changing the logic. It looks like I forgot to put the commit description. 

This was done because I believe the previous code was buggy. However, the change was related to few lines below it, where I changed it to `if ($nbColumns >= $i) {`. 

I think it is safe to revert back to previous syntax, which was a little bit more verbose, but clearer. According to the code, `getFieldsMeta()` can return `?array`. The condition can be `if($meta === null || $meta === [])`. The `|| $meta === []` is absolutely redundant as the code will behave exactly the same without it, so to make it really easy to read, let's keep it as `if($meta === null)`

williamdes commented on libraries/classes/DatabaseInterface.php:
```diff
@@ -1119,7 +1100,7 @@ public function postConnect(): void
         /* Locale for messages */
         $locale = LanguageManager::getInstance()->getCurrentLanguage()->getMySQLLocale();
-        if (! empty($locale)) {
+        if ($locale) {
```
Given the non typed function at https://github.com/phpmyadmin/phpmyadmin/blob/70c43a98633dc475834a0a8d008d96e846dd2ec3/libraries/classes/Language.php#L111

I think it should be explicit
Please try to not widen the possible values that can make if conditions valid :)

kamil-tekiela replied:
I haven't widened any condition. I have tried my best to not change any logic. In this case the condition remains the same. 

`empty()` is equivalent to `! isset($var) || ! $var`. See https://www.php.net/manual/en/function.empty.php
This means that `! empty()` is `! (! isset($var) || ! $var)`, which is `(isset($var) && $var)`.
Since we know that the variable is defined this becomes: `($var !== null && $var)` which can be simply written as `if($var)`. 

I have not widened the condition. The change I made has nothing to do with the type returned by the method. I am happy to narrow it down if you think it's safe, but as I said I tried not to change any logic.

williamdes commented on libraries/classes/DatabaseInterface.php:
```diff
@@ -1594,56 +1556,52 @@ public function getRoutines(
             );
             $result = $this->fetchResult($query);
-            if (! empty($result)) {
+            if ($result) {
```
should we check for false ?

kamil-tekiela replied:
Thanks for pointing this out, This looks wrong indeed. Let me investigate.

kamil-tekiela replied:
Thanks! The `if` statements were redundant. This function always returns an array. I removed them now. 

I then decided to add a return type to `fetchResult()` to catch similar issues. This led me to some Psalm fixes and errors in unit tests. All of this should be fixed in the latest commit.

kamil-tekiela committed:
Widen the condition per williamdes suggestion

This will allow value such as '' and '0' to pass through. Previously
`if(! empty($database))` and `if($database)` dissallowed them.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

kamil-tekiela committed:
fetchResult() should return only arrays

Tests fixed to return arrays. Non-empty in case of true, empty in case
of false.
Fixed a lot of Psalm issues where the condition was always true.

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

mauriciofauth commented:
Merged. Thank you very much!

williamdes reviewed (commented):
:100:
